File: nodejs/src/client/process-intelligence.ts
/**
 * Process Intelligence API: decision traces (LOX-1478), entity context (LOX-1627),
 * causal chain / similar / linked create (LOX-1226).
 *
 * Backend (process-intelligence microservice):
 *   GET  /organizations/:organization_id/decision-traces
 *   POST /organizations/:organization_id/decision-traces
 *   GET  /organizations/:organization_id/decision-traces/:trace_id/chain
 *   GET  /organizations/:organization_id/decision-traces/:trace_id/similar
 *   GET  /organizations/:organization_id/context
 *
 * Thin HTTP wrap only — platform owns graph walk / ranking logic.
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type {
  CreateDecisionTraceInput,
  CreateDecisionTraceResponse,
  CreateDecisionTraceResult,
  DecisionChainResult,
  DecisionTracesListParams,
  DecisionTracesListResponse,
  EntityContextResponse,
  GetDecisionChainParams,
  GetDecisionChainResponse,
  GetEntityContextParams,
  GetSimilarDecisionsParams,
  GetSimilarDecisionsResponse,
  SimilarDecisionsResult,
} from './process-intelligence-types.js';

const BASE = '/process-intelligence';

function buildQuery(params?: DecisionTracesListParams): string {
  if (!params || Object.keys(params).length === 0) return '';
  const searchParams = new URLSearchParams();
  if (params.correlation_key != null) searchParams.set('correlation_key', params.correlation_key);
  if (params.correlation_value != null)
    searchParams.set('correlation_value', params.correlation_value);
  if (params.decision_point != null) searchParams.set('decision_point', params.decision_point);
  if (typeof params.is_exception === 'boolean')
    searchParams.set('is_exception', params.is_exception ? 'true' : 'false');
  if (params.precedent != null) searchParams.set('precedent', params.precedent);
  if (params.page != null) searchParams.set('page', String(params.page));
  if (params.page_size != null) searchParams.set('page_size', String(params.page_size));
  const qs = searchParams.toString();
  return qs ? `?${qs}` : '';
}

function buildChainQuery(params?: GetDecisionChainParams): string {
  if (!params) return '';
  const searchParams = new URLSearchParams();
  if (params.max_depth != null) searchParams.set('max_depth', String(params.max_depth));
  if (params.direction != null) searchParams.set('direction', params.direction);
  const qs = searchParams.toString();
  return qs ? `?${qs}` : '';
}

function buildSimilarQuery(params?: GetSimilarDecisionsParams): string {
  if (!params || params.limit == null) return '';
  return `?limit=${encodeURIComponent(String(params.limit))}`;
}

function tracesBase(organizationId: string): string {
  return `${BASE}/organizations/${encodeURIComponent(organizationId)}/decision-traces`;
}

export type ProcessIntelligenceApi = ReturnType<typeof createProcessIntelligenceApi>;

/**
 * Create the process intelligence API surface:
 * decisionTraces.list / create / getChain / getSimilar, getEntityContext.
 */
export function createProcessIntelligenceApi(http: LoxtepHttpClient): {
  decisionTraces: {
    list: (
      organizationId: string,
      params?: DecisionTracesListParams
    ) => Promise<DecisionTracesListResponse['data']>;
    create: (
      organizationId: string,
      body: CreateDecisionTraceInput
    ) => Promise<CreateDecisionTraceResult>;
    getChain: (
      organizationId: string,
      traceId: string,
      params?: GetDecisionChainParams
    ) => Promise<DecisionChainResult>;
    getSimilar: (
      organizationId: string,
      traceId: string,
      params?: GetSimilarDecisionsParams
    ) => Promise<SimilarDecisionsResult>;
  };
  getEntityContext: (
    organizationId: string,
    params: GetEntityContextParams
  ) => Promise<EntityContextResponse['data']>;
} {
  return {
    decisionTraces: {
      async list(
        organizationId: string,
        params?: DecisionTracesListParams
      ): Promise<DecisionTracesListResponse['data']> {
        const path = `${tracesBase(organizationId)}${buildQuery(params)}`;
        const res = await http.get<DecisionTracesListResponse>(path);
        return res.data;
      },

      async create(
        organizationId: string,
        body: CreateDecisionTraceInput
      ): Promise<CreateDecisionTraceResult> {
        const res = await http.post<CreateDecisionTraceResponse>(tracesBase(organizationId), body);
        return res.data;
      },

      async getChain(
        organizationId: string,
        traceId: string,
        params?: GetDecisionChainParams
      ): Promise<DecisionChainResult> {
        const path = `${tracesBase(organizationId)}/${encodeURIComponent(traceId)}/chain${buildChainQuery(params)}`;
        const res = await http.get<GetDecisionChainResponse>(path);
        return res.data;
      },

      async getSimilar(
        organizationId: string,
        traceId: string,
        params?: GetSimilarDecisionsParams
      ): Promise<SimilarDecisionsResult> {
        const path = `${tracesBase(organizationId)}/${encodeURIComponent(traceId)}/similar${buildSimilarQuery(params)}`;
        const res = await http.get<GetSimilarDecisionsResponse>(path);
        return res.data;
      },
    },
    async getEntityContext(
      organizationId: string,
      params: GetEntityContextParams
    ): Promise<EntityContextResponse['data']> {
      const searchParams = new URLSearchParams();
      searchParams.set('entity_type', params.entity_type);
      searchParams.set('entity_id', params.entity_id);
      const path = `${BASE}/organizations/${encodeURIComponent(organizationId)}/context?${searchParams.toString()}`;
      const res = await http.get<EntityContextResponse>(path);
      return res.data;
    },
  };
}

File: python/src/loxtep/process_intelligence.py
"""
Process Intelligence API (LOX-1478, LOX-1627, LOX-1226).
decision_traces.list / create / get_chain / get_similar, get_entity_context.

Backend:
  GET  /process-intelligence/organizations/:org_id/decision-traces
  POST /process-intelligence/organizations/:org_id/decision-traces
  GET  /process-intelligence/organizations/:org_id/decision-traces/:trace_id/chain
  GET  /process-intelligence/organizations/:org_id/decision-traces/:trace_id/similar
  GET  /process-intelligence/organizations/:org_id/context

Thin HTTP wrap only — platform owns graph walk / ranking logic.
"""

from typing import Any, Optional
from urllib.parse import urlencode

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

BASE = "/process-intelligence"


def _build_decision_traces_qs(params: dict[str, Any]) -> str:
    clean: dict[str, str] = {}
    for k, v in params.items():
        if v is None:
            continue
        if k == "is_exception" and isinstance(v, bool):
            clean[k] = "true" if v else "false"
        else:
            clean[k] = str(v)
    return "?" + urlencode(clean) if clean else ""


def _traces_base(organization_id: str) -> str:
    return f"{BASE}/organizations/{organization_id}/decision-traces"


def _unwrap_data(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


class ProcessIntelligenceApi:
    """Sync process intelligence surface including LOX-1226 chain/similar/create."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def decision_traces_list(
        self,
        organization_id: str,
        *,
        correlation_key: Optional[str] = None,
        correlation_value: Optional[str] = None,
        decision_point: Optional[str] = None,
        is_exception: Optional[bool] = None,
        precedent: Optional[str] = None,
        page: int = 1,
        page_size: int = 20,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if correlation_key is not None:
            params["correlation_key"] = correlation_key
        if correlation_value is not None:
            params["correlation_value"] = correlation_value
        if decision_point is not None:
            params["decision_point"] = decision_point
        if is_exception is not None:
            params["is_exception"] = is_exception
        if precedent is not None:
            params["precedent"] = precedent
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}{qs}"
        return _unwrap_data(self._http.get(path))

    def decision_traces_create(
        self,
        organization_id: str,
        body: dict[str, Any],
    ) -> Any:
        """POST entity-level decision trace (supports links / precedent_id from LOX-1226)."""
        return _unwrap_data(self._http.post(_traces_base(organization_id), body))

    def decision_traces_get_chain(
        self,
        organization_id: str,
        trace_id: str,
        *,
        max_depth: Optional[int] = None,
        direction: Optional[str] = None,
    ) -> Any:
        """GET causal chain for a decision trace (LOX-1226)."""
        params: dict[str, Any] = {}
        if max_depth is not None:
            params["max_depth"] = max_depth
        if direction is not None:
            params["direction"] = direction
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}/{trace_id}/chain{qs}"
        return _unwrap_data(self._http.get(path))

    def decision_traces_get_similar(
        self,
        organization_id: str,
        trace_id: str,
        *,
        limit: Optional[int] = None,
    ) -> Any:
        """GET similar / precedent-ranked decisions (LOX-1226)."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}/{trace_id}/similar{qs}"
        return _unwrap_data(self._http.get(path))

    def get_entity_context(
        self,
        organization_id: str,
        entity_type: str,
        entity_id: str,
    ) -> Any:
        """Get entity context for process intelligence (entity-specific structure)."""
        qs = urlencode({"entity_type": entity_type, "entity_id": entity_id})
        path = f"{BASE}/organizations/{organization_id}/context?{qs}"
        return _unwrap_data(self._http.get(path))


class AsyncProcessIntelligenceApi:
    """Async process intelligence surface including LOX-1226 chain/similar/create."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def decision_traces_list(
        self,
        organization_id: str,
        *,
        correlation_key: Optional[str] = None,
        correlation_value: Optional[str] = None,
        decision_point: Optional[str] = None,
        is_exception: Optional[bool] = None,
        precedent: Optional[str] = None,
        page: int = 1,
        page_size: int = 20,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if correlation_key is not None:
            params["correlation_key"] = correlation_key
        if correlation_value is not None:
            params["correlation_value"] = correlation_value
        if decision_point is not None:
            params["decision_point"] = decision_point
        if is_exception is not None:
            params["is_exception"] = is_exception
        if precedent is not None:
            params["precedent"] = precedent
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}{qs}"
        return _unwrap_data(await self._http.get(path))

    async def decision_traces_create(
        self,
        organization_id: str,
        body: dict[str, Any],
    ) -> Any:
        """POST entity-level decision trace (supports links / precedent_id from LOX-1226)."""
        return _unwrap_data(await self._http.post(_traces_base(organization_id), body))

    async def decision_traces_get_chain(
        self,
        organization_id: str,
        trace_id: str,
        *,
        max_depth: Optional[int] = None,
        direction: Optional[str] = None,
    ) -> Any:
        """GET causal chain for a decision trace (LOX-1226)."""
        params: dict[str, Any] = {}
        if max_depth is not None:
            params["max_depth"] = max_depth
        if direction is not None:
            params["direction"] = direction
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}/{trace_id}/chain{qs}"
        return _unwrap_data(await self._http.get(path))

    async def decision_traces_get_similar(
        self,
        organization_id: str,
        trace_id: str,
        *,
        limit: Optional[int] = None,
    ) -> Any:
        """GET similar / precedent-ranked decisions (LOX-1226)."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        qs = _build_decision_traces_qs(params)
        path = f"{_traces_base(organization_id)}/{trace_id}/similar{qs}"
        return _unwrap_data(await self._http.get(path))

    async def get_entity_context(
        self,
        organization_id: str,
        entity_type: str,
        entity_id: str,
    ) -> Any:
        """Get entity context for process intelligence (entity-specific structure)."""
        qs = urlencode({"entity_type": entity_type, "entity_id": entity_id})
        path = f"{BASE}/organizations/{organization_id}/context?{qs}"
        return _unwrap_data(await self._http.get(path))
